Question: Is it possible to add "fixed" request data that are always run (esp. with cli)?

Hi, thanks a lot for this really cool library.

(running schemathesis 2.3.3 on Python 3.6.5)

I'm having difficulties grasp if I can somehow add "fixed" data to the tests that is always run, in addition to the strategy-generated random data. I want this to ensure some valid interactions are performed against the API, e.g. actually getting a resource with a valid resource id, so that a response can get validated.

While I understand that maybe this should rather be accomplished with accompanying tests (?) of another notion (I'm new to
property-based testing) it is just too tempting to make use of schemathesis capabilities for it... ;-)

https://schemathesis.readthedocs.io/en/stable/faq.html#how-can-i-use-database-objects-ids-in-tests hints in the right direction of setting some valid ID but I'm having a hard time mapping this to cli mode.

I think a have a (kinda) working hook-based solution using along these lines:

```
# org_schemathesis.py
from typing import List
                                                                               
import schemathesis
from schemathesis.models import Case
from hypothesis import strategies
 
 
@schemathesis.hooks.register
def before_add_examples(
        context: schemathesis.hooks.HookContext,
        examples: List[Case],
    ) -> None:
    examples.append(
        Case(endpoint=context.endpoint, path_parameters={'employeeId': 12345})
        )
```

If I run this like
```
rm -fr ~/.hypothesis ./cassette.yaml && schemathesis --pre-run org_schemathesis run org-service_0.0.2.yaml --base-url=http://my.server.net:8081 --show-errors-tracebacks --auth="$ORG_AUTH" --checks=response_schema_conformance  -E /employees/ --store-network-log=cassette.yaml
```
I can see my hardcoded `employeeId` put to use in the stored cassette.yaml, along with an actual response, with a 
status: 'SUCCESS' result.

Is this a proper way to do this? IIUC the example data will always get used for the tests.

Another trouble, slightly related: If I run above with `--checks=all` my hardcoded test data is still reported and still has a a proper
response status & data, but it is mysteriously marked as a 'FAILURE' now. 
I suspect this is due to the fact that I now see other failures (for status_code_conformance checks) ; I fear there's some bug in
storing the proper status for each `http_interaction ` in `cassette.yaml`.

Best regards,
Holger

Hi Dmitry,

by final status, do you mean that if a single request GET /endpoint/<param> is performed, than this will get checked for

- not_a_server_error
- status_code_conformance
- content_type_conformance
- response_schema_conformance

and if any these checks for one single endpoint invocation fail then it is status: 'FAILURE'?

Makes sense. Yes, I suppose #702 would then provide helpful insight into the failure reason.

But I think I have a slightly different problem: 

I force-add a fixed example to the test samples by pre-run hook, e.g. to make sure certain known values get "triggered". I run with `--checks=response_schema_conformance` i.e. exclude `status_code_conformance` checking (amongst others). 
I run this against a service with status conformance problems - all tests are successful, as these errors aren't checked for.
I then rerun all checks and expect some tests to fail, but I'd also expect some tests to be successful.

Minimal example using fastapi:

A simple fastapi hello-world-service. The marked lines have been added after an initial retrieval of its openapi.json:
```
# main.py
from fastapi import FastAPI, HTTPException
from pydantic import BaseModel


app = FastAPI(title='hello-world-service')


# data model
class HelloResponse(BaseModel):
    message: str


@app.get("/hello", response_model=HelloResponse)
def read_hello():
    return {"message": "Hello, world!"}


@app.get("/hello/{name}", response_model=HelloResponse)
def read_hello_name(name: str):
    ###### 
    # These lines have been added after retrieving openapi.json from fastapi service,
    # so they provoke a status result that is definitely not documented in our oas spec.
    if name == '0':
        raise HTTPException(status_code=404, detail="not found")
    ###### 
    return {"message": f"Hello, {name}!"}
```

My hook:
```
# add_examples.py
from typing import List

import schemathesis
from schemathesis.models import Case


@schemathesis.hooks.register
def before_add_examples(
        context: schemathesis.hooks.HookContext,
        examples: List[Case],
    ) -> None:
    examples.append(
        Case(endpoint=context.endpoint, path_parameters={'name': 'holger'})
        )

```

Run test subset:
```
rm -fr ~/.hypothesis ./cassette.yaml && ~/pydev/venv/schemathesis-venv/bin/schemathesis --pre-run add_examples run openapi.json --base-url=http://myserver.net:8000 --show-errors-tracebacks --checks=response_schema_conformance  -E /hello/ --store-network-log=cassette.yaml
======================= Schemathesis test session starts ======================
platform Linux -- Python 3.6.5, schemathesis-2.3.3, hypothesis-5.23.3, hypothesis_jsonschema-0.17.3, jsonschema-3.2.0
rootdir: ...
Specification version: Open API 3.0.2
Workers: 1
collected endpoints: 1

GET /hello/{name} .                                                     [100%]

=================================== SUMMARY ===================================

Performed checks:
    response_schema_conformance                    101 / 101 passed          PASSED 

Network log: cassette.yaml
```

All tests are reported as succesful, including my custom added example:
```
$ grep "status: 'SUCCESS'" cassette.yaml |wc -l
101
```

Now rerun with --checks=all:

```
$ rm -fr ~/.hypothesis ./cassette.yaml && ~/pydev/venv/schemathesis-venv/bin/schemathesis --pre-run add_examples run openapi.json --base-url=http://myserver.net:8000 --show-errors-tracebacks --checks=all  -E /hello/ --store-network-log=cassette.yaml
======================= Schemathesis test session starts ======================
platform Linux -- Python 3.6.5, schemathesis-2.3.3, hypothesis-5.23.3, hypothesis_jsonschema-0.17.3, jsonschema-3.2.0
rootdir: ...
Specification version: Open API 3.0.2
Workers: 1
collected endpoints: 1

GET /hello/{name} F                                                     [100%]

=================================== FAILURES ==================================
______________________________ GET: /hello/{name} _____________________________
1. Received a response with a status code, which is not defined in the schema: 404

Declared status codes: 200, 422

Check           : status_code_conformance
Path parameters : {'name': '0'}

Run this Python code to reproduce this failure: 

    requests.get('http://myserver.net:8000/hello/0')

Or add this option to your command line parameters: --hypothesis-seed=87345815337462284460351829168824299911
=================================== SUMMARY ===================================

Performed checks:
    not_a_server_error                             4 / 4 passed          PASSED 
    status_code_conformance                        1 / 4 passed          FAILED 
    content_type_conformance                       4 / 4 passed          PASSED 
    response_schema_conformance                    4 / 4 passed          PASSED 

Network log: cassette.yaml
```

Now, all tests are reported as failures; this includes my custom fixed example:

```
command: 'schemathesis --pre-run add_examples run openapi.json --base-url=http://myserver.net:8000 --show-errors-tracebacks --checks=all -E /hello/ --store-network-log=cassette.yaml'
recorded_with: 'Schemathesis 2.3.3'
http_interactions:
- id: '1'
  status: 'FAILURE'
  seed: '138515993875395522801610595088206829211'
  elapsed: '0.006496'
  recorded_at: '2020-10-02T18:41:46.480052'
  request:
    uri: 'http://myserver.net:8000/hello/holger'
    method: 'GET'
    headers:
      User-Agent:
      - 'schemathesis/2.3.3'
      Accept-Encoding:
      - 'gzip, deflate'
      Accept:
      - '*/*'
      Connection:
      - 'keep-alive'
    body:
      encoding: 'utf-8'
      base64_string: ''
  response:
    status:
      code: '200'
      message: 'OK'
    headers:
      date:
      - 'Fri, 02 Oct 2020 16:41:45 GMT'
      server:
      - 'uvicorn'
      content-length:
      - '28'
      content-type:
      - 'application/json'
    body:
      encoding: 'utf8'
      base64_string: 'eyJtZXNzYWdlIjoiSGVsbG8sIGhvbGdlciEifQ=='
    http_version: '1.1'
...
```

I'd say this clearly should still be 'SUCCESS'. Of course, some of the other tests are bound to fail :-).

Hm, on a sidenote: Why does schemathesis run only 4 tests in the latter case? Something to do with hypothesis/property based testing works generally?

Best regards,
Holger

--- a/docs/changelog.rst
+++ b/docs/changelog.rst
@@ -16,6 +16,8 @@ Changelog
 **Fixed**
 
 - Default ``User-Agent`` header in ``Case.call``. `#717`_
+- Status of individual interactions in VCR cassettes. Before this change, all statuses were taken from the overall test outcome,
+  rather than from the check results for a particular response. `#695`_
 
 **Changed**
 
@@ -1387,6 +1389,7 @@ Deprecated
 .. _#708: https://github.com/schemathesis/schemathesis/issues/708
 .. _#705: https://github.com/schemathesis/schemathesis/issues/705
 .. _#702: https://github.com/schemathesis/schemathesis/issues/702
+.. _#695: https://github.com/schemathesis/schemathesis/issues/695
 .. _#692: https://github.com/schemathesis/schemathesis/issues/692
 .. _#686: https://github.com/schemathesis/schemathesis/issues/686
 .. _#684: https://github.com/schemathesis/schemathesis/issues/684

--- a/docs/cli.rst
+++ b/docs/cli.rst
@@ -260,7 +260,7 @@ Schemathesis provides the following extra fields:
 
 - ``command``. Full CLI command used to run Schemathesis.
 - ``http_interactions.id``. A numeric interaction ID within the current cassette.
-- ``http_interactions.status``. Type of test outcome is one of ``SUCCESS``, ``FAILURE``, ``ERROR``.
+- ``http_interactions.status``. Type of test outcome is one of ``SUCCESS``, ``FAILURE``. The status value is calculated from individual checks statuses - if any check failed, then the final status is ``FAILURE``.
 - ``http_interactions.seed``. The Hypothesis seed used in that particular case could be used as an argument to ``--hypothesis-seed`` CLI option to reproduce this request.
 - ``http_interactions.elapsed``. Time in seconds that a request took.
 - ``http_interactions.checks``. A list of executed checks and and their status.
@@ -288,7 +288,7 @@ Saved cassettes can be replayed with ``schemathesis replay`` command. Additional
 replay by these parameters:
 
 - ``id``. Specific, unique ID;
-- ``status``. Replay only interactions with this status (``SUCCESS``, ``FAILURE`` or ``ERROR``);
+- ``status``. Replay only interactions with this status (``SUCCESS`` or ``FAILURE``);
 - ``uri``. A regular expression for request URI;
 - ``method``. A regular expression for request method;
 

--- a/src/schemathesis/cli/cassettes.py
+++ b/src/schemathesis/cli/cassettes.py
@@ -48,7 +48,6 @@ class CassetteWriter(EventHandler):
             seed = cast(int, event.result.seed)
             self.queue.put(
                 Process(
-                    status=event.status.name.upper(),
                     seed=seed,
                     interactions=event.result.interactions,
                     checks=event.result.checks,
@@ -74,7 +73,6 @@ class Initialize:
 class Process:
     """A new chunk of data should be processed."""
 
-    status: str = attr.ib()  # pragma: no mutate
     seed: int = attr.ib()  # pragma: no mutate
     interactions: List[Interaction] = attr.ib()  # pragma: no mutate
     checks: List[SerializedCheck] = attr.ib()  # pragma: no mutate
@@ -132,9 +130,10 @@ http_interactions:"""
             )
         elif isinstance(item, Process):
             for interaction in item.interactions:
+                status = interaction.status.name.upper()
                 stream.write(
                     f"""\n- id: '{current_id}'
-  status: '{item.status}'
+  status: '{status}'
   seed: '{item.seed}'
   elapsed: '{interaction.response.elapsed}'
   recorded_at: '{interaction.recorded_at}'

--- a/src/schemathesis/models.py
+++ b/src/schemathesis/models.py
@@ -542,17 +542,26 @@ class Interaction:
 
     request: Request = attr.ib()  # pragma: no mutate
     response: Response = attr.ib()  # pragma: no mutate
+    status: Status = attr.ib()  # pragma: no mutate
     recorded_at: str = attr.ib(factory=lambda: datetime.datetime.now().isoformat())  # pragma: no mutate
 
     @classmethod
-    def from_requests(cls, response: requests.Response) -> "Interaction":
-        return cls(request=Request.from_prepared_request(response.request), response=Response.from_requests(response))
+    def from_requests(cls, response: requests.Response, status: Status) -> "Interaction":
+        return cls(
+            request=Request.from_prepared_request(response.request),
+            response=Response.from_requests(response),
+            status=status,
+        )
 
     @classmethod
-    def from_wsgi(cls, case: Case, response: WSGIResponse, headers: Dict[str, Any], elapsed: float) -> "Interaction":
+    def from_wsgi(  # pylint: disable=too-many-arguments
+        cls, case: Case, response: WSGIResponse, headers: Dict[str, Any], elapsed: float, status: Status
+    ) -> "Interaction":
         session = requests.Session()
         session.headers.update(headers)
-        return cls(request=Request.from_case(case, session), response=Response.from_wsgi(response, elapsed))
+        return cls(
+            request=Request.from_case(case, session), response=Response.from_wsgi(response, elapsed), status=status
+        )
 
 
 @attr.s(slots=True, repr=False)  # pragma: no mutate
@@ -593,11 +602,13 @@ class TestResult:
     def add_error(self, exception: Exception, example: Optional[Case] = None) -> None:
         self.errors.append((exception, example))
 
-    def store_requests_response(self, response: requests.Response) -> None:
-        self.interactions.append(Interaction.from_requests(response))
+    def store_requests_response(self, response: requests.Response, status: Status) -> None:
+        self.interactions.append(Interaction.from_requests(response, status))
 
-    def store_wsgi_response(self, case: Case, response: WSGIResponse, headers: Dict[str, Any], elapsed: float) -> None:
-        self.interactions.append(Interaction.from_wsgi(case, response, headers, elapsed))
+    def store_wsgi_response(  # pylint: disable=too-many-arguments
+        self, case: Case, response: WSGIResponse, headers: Dict[str, Any], elapsed: float, status: Status
+    ) -> None:
+        self.interactions.append(Interaction.from_wsgi(case, response, headers, elapsed, status))
 
 
 @attr.s(slots=True, repr=False)  # pragma: no mutate

--- a/src/schemathesis/runner/impl/core.py
+++ b/src/schemathesis/runner/impl/core.py
@@ -284,9 +284,15 @@ def _network_test(
     response = case.call(session=session, headers=headers, timeout=timeout)
     context = TargetContext(case=case, response=response, response_time=response.elapsed.total_seconds())
     run_targets(targets, context)
-    if store_interactions:
-        result.store_requests_response(response)
-    run_checks(case, checks, result, response, context.response_time * 1000, max_response_time)
+    status = Status.success
+    try:
+        run_checks(case, checks, result, response, context.response_time * 1000, max_response_time)
+    except CheckFailed:
+        status = Status.failure
+        raise
+    finally:
+        if store_interactions:
+            result.store_requests_response(response, status)
     feedback.add_test_case(case, response)
     return response
 
@@ -344,10 +350,16 @@ def _wsgi_test(
         elapsed = time.monotonic() - start
     context = TargetContext(case=case, response=response, response_time=elapsed)
     run_targets(targets, context)
-    if store_interactions:
-        result.store_wsgi_response(case, response, headers, elapsed)
     result.logs.extend(recorded.records)
-    run_checks(case, checks, result, response, context.response_time * 1000, max_response_time)
+    status = Status.success
+    try:
+        run_checks(case, checks, result, response, context.response_time * 1000, max_response_time)
+    except CheckFailed:
+        status = Status.failure
+        raise
+    finally:
+        if store_interactions:
+            result.store_wsgi_response(case, response, headers, elapsed, status)
     feedback.add_test_case(case, response)
     return response
 
@@ -405,8 +417,14 @@ def _asgi_test(
     response = case.call_asgi(headers=headers)
     context = TargetContext(case=case, response=response, response_time=response.elapsed.total_seconds())
     run_targets(targets, context)
-    if store_interactions:
-        result.store_requests_response(response)
-    run_checks(case, checks, result, response, context.response_time * 1000, max_response_time)
+    status = Status.success
+    try:
+        run_checks(case, checks, result, response, context.response_time * 1000, max_response_time)
+    except CheckFailed:
+        status = Status.failure
+        raise
+    finally:
+        if store_interactions:
+            result.store_requests_response(response, status)
     feedback.add_test_case(case, response)
     return response

--- a/test/cli/test_cassettes.py
+++ b/test/cli/test_cassettes.py
@@ -22,6 +22,10 @@ def load_cassette(path):
         return yaml.safe_load(fd)
 
 
+def load_response_body(cassette, idx):
+    return base64.b64decode(cassette["http_interactions"][idx]["response"]["body"]["base64_string"])
+
+
 @pytest.mark.endpoints("success", "upload_file")
 def test_store_cassette(cli, schema_url, cassette_path):
     result = cli.run(
@@ -35,13 +39,35 @@ def test_store_cassette(cli, schema_url, cassette_path):
     assert cassette["http_interactions"][0]["status"] == "SUCCESS"
     assert cassette["http_interactions"][0]["seed"] == "1"
     assert float(cassette["http_interactions"][0]["elapsed"]) >= 0
-    data = base64.b64decode(cassette["http_interactions"][0]["response"]["body"]["base64_string"])
-    assert data == b'{"success": true}'
+    assert load_response_body(cassette, 0) == b'{"success": true}'
     assert all("checks" in interaction for interaction in cassette["http_interactions"])
     assert len(cassette["http_interactions"][0]["checks"]) == 1
     assert cassette["http_interactions"][0]["checks"][0]["name"] == "not_a_server_error"
     assert cassette["http_interactions"][0]["checks"][0]["status"] == "SUCCESS"
     assert cassette["http_interactions"][0]["checks"][0]["message"] is None
+
+
+@pytest.mark.endpoints("flaky")
+def test_interaction_status(cli, openapi3_schema_url, cassette_path):
+    # See GH-695
+    # When an endpoint has responses with SUCCESS and FAILURE statuses
+    result = cli.run(
+        openapi3_schema_url,
+        f"--store-network-log={cassette_path}",
+        "--hypothesis-max-examples=5",
+        "--hypothesis-seed=1",
+    )
+    assert result.exit_code == ExitCode.TESTS_FAILED, result.stdout
+    cassette = load_cassette(cassette_path)
+    assert len(cassette["http_interactions"]) == 3
+    # Then their statuses should be reflected in the "status" field
+    # And it should not be overridden by the overall test status
+    assert cassette["http_interactions"][0]["status"] == "FAILURE"
+    assert load_response_body(cassette, 0) == b"500: Internal Server Error"
+    assert cassette["http_interactions"][1]["status"] == "SUCCESS"
+    assert load_response_body(cassette, 1) == b'{"result": "flaky!"}'
+    assert cassette["http_interactions"][2]["status"] == "SUCCESS"
+    assert load_response_body(cassette, 2) == b'{"result": "flaky!"}'
 
 
 def test_encoding_error(testdir, cli, cassette_path, openapi3_base_url):

